keys/keybase.go

```diff
@@ -32,58 +26,59 @@ func New(db dbm.DB, codec words.Codec) dbKeybase {
 
 var _ Keybase = dbKeybase{}
 
-// Create generates a new key and persists it storage, encrypted using the passphrase.
-// It returns the generated seedphrase (mnemonic) and the key Info.
-// It returns an error if it fails to generate a key for the given algo type,
-// or if another key is already stored under the same name.
-func (kb dbKeybase) Create(name, passphrase, algo string) (string, Info, error) {
+// Create generates a new key and persists it to storage, encrypted
+// using the passphrase.  It returns the generated seedphrase
+// (mnemonic) and the key Info.  It returns an error if it fails to
+// generate a key for the given algo type, or if another key is
+// already stored under the same name.
+func (kb dbKeybase) Create(name, passphrase string, algo CryptoAlgo) (Info, string, error) {
 	// NOTE: secret is SHA256 hashed by secp256k1 and ed25519.
 	// 16 byte secret corresponds to 12 BIP39 words.
 	// XXX: Ledgers use 24 words now - should we ?
 	secret := crypto.CRandBytes(16)
-	key, err := generate(algo, secret)
+	priv, err := generate(algo, secret)
 	if err != nil {
-		return "", Info{}, err
+		return Info{}, "", err
 	}
 
 	// encrypt and persist the key
-	public := kb.writeKey(key, name, passphrase)
+	info := kb.writeKey(priv, name, passphrase)
+
+	// we append the type byte to the serialized secret to help with
+	// recovery
+	// ie [secret] = [type] + [secret]
+	typ := cryptoAlgoToByte(algo)
+	secret = append([]byte{typ}, secret...)
 
 	// return the mnemonic phrase
 	words, err := kb.codec.BytesToWords(secret)
-	seedphrase := strings.Join(words, " ")
-	return seedphrase, public, err
-}
-
-// Recover converts a seedphrase to a private key and persists it, encrypted with the given passphrase.
-// Functions like Create, but seedphrase is input not output.
-func (kb dbKeybase) Recover(name, passphrase, algo string, seedphrase string) (Info, error) {
-
-	key, err := kb.SeedToPrivKey(algo, seedphrase)
-	if err != nil {
-		return Info{}, err
-	}
-
-	// Valid seedphrase. Encrypt key and persist to disk.
-	public := kb.writeKey(key, name, passphrase)
-	return public, nil
+	seed := strings.Join(words, " ")
+	return info, seed, err
 }
 
-// SeedToPrivKey returns the private key corresponding to a seedphrase
-// without persisting the private key.
-// TODO: enable the keybase to just hold these in memory so we can sign without persisting (?)
-func (kb dbKeybase) SeedToPrivKey(algo, seedphrase string) (crypto.PrivKey, error) {
```

Comment:
What if we just want to use a private key from seed without persisting it to disk? that's what this function is for ...

Reply:
who calls it?  I couldn't find any usage needs.

Reply:
We haven't exposed it yet but we should. Also the UI is now asking for it (https://github.com/cosmos/cosmos-sdk/issues/301) . We defintely should have a way to enter a seed and use the key without it being persisted to disk.